libraries/adaptive-expressions/src/functionUtils.ts

```diff
@@ -0,0 +1,1045 @@
+     * Transform C# period and unit into js period and unit
+     * @param duration C# duration
+     * @param cSharpStr C# unit.
+     */
+    public static timeUnitTransformer(duration: number, cSharpStr: string): { duration: number; tsStr: string } {
+        switch (cSharpStr) {
+            case 'Day': return { duration, tsStr: 'days' };
+            case 'Week': return { duration: duration * 7, tsStr: 'days' };
+            case 'Second': return { duration, tsStr: 'seconds' };
+            case 'Minute': return { duration, tsStr: 'minutes' };
+            case 'Hour': return { duration, tsStr: 'hours' };
+            case 'Month': return { duration, tsStr: 'months' };
+            case 'Year': return { duration, tsStr: 'years' };
+            default: return { duration, tsStr: undefined };
+        }
+    }
+
+    /**
+     * Parse timex funcition.
+     * @param timexExpr String or TimexProperty input.
+     * @returns TimexProperty and error.
+     */
+    public static parseTimexProperty(timexExpr: any): { timexProperty: TimexProperty; error: string } {
+        let parsed: TimexProperty;
+        if (timexExpr instanceof TimexProperty) {
+            parsed = timexExpr;
+        } else if (typeof timexExpr === 'string') {
+            parsed = new TimexProperty(timexExpr);
+        } else {
+            parsed = new TimexProperty(timexExpr);
+            if (parsed === undefined || Object.keys(parsed).length === 0) {
+                return { timexProperty: parsed, error: `${timexExpr} requires a TimexProperty or a string as a argument` };
+            }
+        }
+
+        return { timexProperty: parsed, error: undefined };
+    }
+
+    /**
+     * Wrap string or undefined into string. Default to empty string.
+     * @param input input string
+     */
+    public static parseStringOrUndefined(input: string | undefined): string {
+        if (typeof input === 'string') {
+            return input;
+        } else {
+            return '';
+        }
+    }
+
+    /**
+     * Try to accumulate the path from an Accessor or Element, from right to left.
+     * @param expression expression.
+     * @param state scope.
+     * @param options Options used in evaluation.
+     * @returns Return the accumulated path and the expression left unable to accumulate.
+     */
+    public static tryAccumulatePath(expression: Expression, state: MemoryInterface, options: Options): { path: string; left: any; error: string } {
+        let path = '';
+        let left = expression;
+        while (left !== undefined) {
+            if (left.type === ExpressionType.Accessor) {
+                path = (left.children[0] as Constant).value + '.' + path;
+                left = left.children.length === 2 ? left.children[1] : undefined;
+            } else if (left.type === ExpressionType.Element) {
+                let value: any;
+                let error: string;
+                ({ value, error } = left.children[1].tryEvaluate(state, options));
+
```

Comment:
This would be much clearer as:
  let {value, error} = left.children[1].tryEvaluate(state, options)
It is weird to do the let and then the () assignment.